impute_sex, now 100% python

danking opened the pull request:
cc: @tpoterba I added some new type check stuff

I added `round` `ceil` and `floor` as I was writing tests and then I didn't need them. They still seem useful, so I included them.

I had to add a tolerance parameter to `Table.same` and my implementation is kind of ugly and bad but I can't think of an obviously better way to do this.

Example usages:

```python
imputed_sex = methods.impute_sex(ds.v.locus(), ds.GT)
```

```python
imputed_sex = methods.impute_sex(ds.v.locus(), 
                                 ds.GT,
                                 aaf=gnomad.AF)
```

danking commented:
I actually quite like this with `tablify`. [Take a look at the differences before/after `tablify`](https://github.com/hail-is/hail/pull/2852/commits/31c48bfb8417279b28aa738dab6b56deb0c0b1e2).

tpoterba commented on python/hail/expr/functions.py:
```diff
@@ -515,6 +515,21 @@ def fisher_exact_test(c1, c2, c3, c4):
 
 
+@typecheck(x=oneof(expr_float32, expr_float64))
+def floor(x):
```
why don't these accept integer types and do implicit conversion?
```
In [3]: np.floor(2)
Out[3]: 2.0
```

tpoterba replied:
I think there's pretty much nowhere we should accept floats and not ints.

danking replied:
Doesn't it seem like a mistake to call floor on an integer? I can only think of things like floor'ing the return value of an if, but in that case the int will be promoted by the `if` before the `floor` sees it.

tpoterba replied:
I don't think anything in numpy that expects a float would error on int. Shouldn't we do the same?

danking commented:
@tpoterba back to you, I switched to `expr_numeric` and removed `round` (since that seems controversial now).

We need a final decision about using either:

 - `annotate_all`
 - `tablify`
 - neither, always specify fields as strings

jigold commented on src/main/scala/is/hail/table/Table.scala:
```diff
@@ -158,6 +158,26 @@ object Table {
+      var matched = false
+      var j = 0
+      while (!matched && j < l.length && !used(j)) {
+        matched = t.valuesSimilar(li, r(j), tolerance)
```
I don't understand why you are doing a product comparison here. Shouldn't it be a zip operation?

danking replied:
IIUC, the array of rows here are all the values that had the same key in the left table and the right table, respectively.

So I don't know ahead of time the right ordering of the values on the left and the values on the right such that they're similar. For example, if the left had 1.0 and 2.0 and right had 1.9 and 0.9, and the tolerance was 1.0, I'd want to match the 1.0 with the 0.9 and the and 1.9 with the 2.0. But to find that particular set of pairs, I, in general, need to check all possible set of pairs.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -1144,6 +1287,11 @@ def ld_matrix(dataset, force_local=False):
 
+    .. testsetup::
+
+        dataset = vds.annotate_samples_expr('sa = drop(sa, qc)').to_hail2()
```
Why is this here?

danking replied:
must be a rebase screw up, removed

jigold commented on python/hail/expr/functions.py:
```diff
@@ -577,6 +577,16 @@ def fisher_exact_test(c1, c2, c3, c4):
+@typecheck(x=expr_numeric)
+def floor(x):
+    """The largest integral value that is less than or equal to `x`."""
+    return _func("floor", x.dtype, x)
```
Add doctest examples and Return type.

danking replied:
👍

jigold commented on python/hail/__init__.py:
```diff
@@ -41,6 +42,7 @@
            'methods',
            'eval_expr',
            'eval_expr_typed',
+           'tablify',
```
This is the wrong module. Also, I don't think this should be in the hail or hail2 package.

danking replied:
I don't understand. I figured `tablify` would be generally useful to library writers.

danking commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+
+    Remove samples where imputed sex does not equal reported sex:
+
+    >>> imputed_sex = methods.impute_sex(ds.rows_table().v.locus(), ds.GT)
```
fixed

danking commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+
+    The returned column-key indexed :class:`.Table` has the following fields:
+
+    - **s** -- The column key of `call`
```
definitely. to make the code below do this, I just need a way to refer to the column keys. I think tim is gonna add such functionality when he adds multiple column keys

tpoterba replied:
they exist!

tpoterba replied:
row keys coming in hot, but col keys exist now. `ds.col_key` is the list of strs

danking replied:
ok I think I fixed this to support multiple column keys now

danking commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+    - **is_female** (:class:`.TBoolean`) -- True if the imputed sex is female, false if male, missing if undetermined
+    - **Fstat** (:class:`.TFloat64`) -- Inbreeding coefficient
+    - **nTotal** (:class:`.TInt64`) -- Total number of variants considered
+    - **nCalled**  (:class:`.TInt64`) -- Number of variants with a genotype call
```
👍

danking commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+
+    - **s** -- The column key of `call`
+    - **is_female** (:class:`.TBoolean`) -- True if the imputed sex is female, false if male, missing if undetermined
+    - **Fstat** (:class:`.TFloat64`) -- Inbreeding coefficient
```
👍

danking commented:
I don't like
```python
ds = ds.annotate_rows(locus2 = crosswalk[ds.rsid].locus)
ds = ds.annotate_rows(AF = gnomad[ds.locus2].AF)
sex = methods.impute_sex(ds.locus2, ds.GT, ds.AF)
```
because it's three lines compared to
```python
sex = methods.impute_sex(crosswalk.locus, ds.GT, aaf=gnomad.AF)
```

That's really the only reason. If everyone thinks this is too complicated, I guess it's fine to go with something else, but it seems like a succinct way to describe the computation.

danking commented:
I guess if we can think of better documentation, then the current implementation works with either of those expressions.

danking commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+
+    1. Filter the dataset to loci on the X contig defined by `gr`.
+
+    4. Calculate alternate allele frequency (AAF) for each row from the dataset.
```
foxed

danking commented:
Is "singly-indexed" expression better than "Table-like expression"?

danking commented on python/hail/expr/functions.py:
```diff
@@ -577,6 +577,40 @@ def fisher_exact_test(c1, c2, c3, c4):
+    -------
+    :obj:`.Float64Expression`
+    """
+    return _func("floor", x.dtype, x)
```
Now I'm back to thinking this really should be `expr_float`, how do I specify the return type correctly without duplicating the promotion logic?

tpoterba replied:
`ret_type = unify_types(x.dtype, TFloat32())`

tpoterba replied:
should give you int32/64 -> float32, float64 ->float64

danking commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +104,148 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+           female_threshold=numeric,
+           male_threshold=numeric,
+           aaf=nullable(expr_numeric))
+def impute_sex(locus, call, aaf_threshold=0.0, include_par=False, female_threshold=0.2, male_threshold=0.8, aaf=None):
```
This method works perfectly fine on a locus-only dataset, why impose an unnecessary pre-condition?

tpoterba replied:
we have a whole suite of genetics methods which expect a certain structure for the data. This is how we keep some of our high level of abstraction, while still being totally general. We get a lot more noisy and less user-friendly if we suddenly start requiring the user to pass the locus/alleles field into every method that wants them -- the implementation here is also inconsistent with the other genetics methods like `export_vcf` or `sample_qc`

tpoterba replied:
if you think we're doing the wrong thing, we should discuss. But everything should certainly be consistent either way

cseed commented:
I wanted to chime in on this briefly because I think it a good example use case and its design will influence many future methods, so it is important to get the design right.  Thoughts:

 - the underscore stuff is a non-starter in my opinion, and too clever by half.  A lot of my feedback on your stuff is guided by the general heuristic that you should start by writing down the code you want, and then decide how to implement.  You'd never want to write this _ stuff if you didn't have to.

 - I'm still not quite sure what tablify does (in part because the name is too clever by half and in part because it doesn't appear to always return tables).

 - But I think the idea of tablify is something we want, which is to convert (possibly indexed expressions) back into relational objects (Table, MatrixTable) because the latter support a wider set of operations and don't have the "source mismatch problem".  Tim and I discussed this yesterday and we suggest the following interface:

```
  t = build_table()
    .set_globals(x = 5, batch = batch)
    .set_rows(locus = locus, aaf = aaf)
    .build()
```

and

```
  mt = build_table_matix()
    .set_globals(dataset = dataset)
    .set_rows(locus = locus, aaf = aaf)
    .set_entries(GT = GT)
    .build()
```

where the input expressions for each part must all come from the same source (or be compatible, e.g., constants) and the resulting (matrix) table inherits the keys from the original table.  I think there is an unresolved question about how to handle potential name conflicts (e.g. a column key named locus).

danking commented:
@tpoterba @jigold back to you

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+    call : :class:`.CallExpression`
+        A genotype call for each row and column. The source dataset's row keys
+        must be [[locus], alleles] with types :class:`.TLocus` and
+        :class:`.ArrayStringExpression`. Moreover, the keys must all be biallelic.
```
I think it would be better to say the length of the row key alleles must be equal to 2 (biallelic).

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+    ------
+    :class:`.Table`
+        Sex imputation statistics per sample.
+
```
delete empty line.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+    if (aaf is None):
+        ds = ds.annotate_rows(
+            aaf=(agg.sum(ds.call.num_alt_alleles()).to_float64() /
+                 agg.count_where(f.is_defined(ds.call)) / 2))
```
should this be `* 2` rather than `/ 2`?

danking replied:
divison is left associative so no, but I rewrote it to
```
(... /
 (2.0 * ...))
```
so hopefully that's more clear

tpoterba replied:
can we use call stats?

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+        Sex imputation statistics per sample.
+
+    """
+
```
assert aaf_threshold between [0, 1].

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+
+    The returned column-key indexed :class:`.Table` has the following fields:
+
+    - **s** -- The column key of `call`.
```
I'm not sure how to fix this, but `s` is not always the name of the field. I think we should just delete `**s** --` and it should just be `The column key(s) of call`.

jigold commented on python/hail/methods/tests.py:
```diff
@@ -82,6 +82,47 @@ def compare(ds, min=None, max=None):
+            out = sp.check_output(
+                ["plink", "--vcf", vcf_file, "--const-fid", "--check-sex",
+                 "--silent", "--out", out_file],
+                stderr=sp.STDOUT)
```
Can you add a test using the alternate allele frequency expression computed from call_stats.AF and make sure your results agree with your formula with aaf=None?

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -179,7 +179,8 @@ def impute_sex(call, aaf_threshold=0.0, include_par=False, female_threshold=0.2,
         A genotype call for each row and column. The source dataset's row keys
         must be [[locus], alleles] with types :class:`.TLocus` and
-        :class:`.ArrayStringExpression`. Moreover, the keys must all be biallelic.
+        :class:`.ArrayStringExpression`. Moreover, the alleles array have
```
alleles array must have exactly two elements (biallelic).

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -212,9 +213,7 @@ def locus(ds):
-        ds = ds.annotate_rows(
-            aaf=(agg.sum(ds.call.num_alt_alleles()).to_float64() /
-                 agg.count_where(f.is_defined(ds.call)) / 2))
+        ds = ds.annotate_rows(aaf=agg.call_stats(ds.call, ds.v).AF[0])
```
I think this should be `AF[1]`.

jigold commented on python/hail/methods/tests.py:
```diff
@@ -123,6 +123,10 @@ def test_impute_sex_same_as_plink(self):
 
         self.assertTrue(plink_sex._same(sex, tolerance=1e-3))
 
+        ds = ds.annotate_rows(aaf=(agg.call_stats(ds.GT, ds.v)).AF[0])
```
same -- AF[1]. I double checked and the answer won't change if you use AF[0] or AF[1] because it's using `p(1-p)`. However, I'd prefer to be consistent if the parameter name is alternate allele frequency.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -196,8 +197,8 @@ def impute_sex(call, aaf_threshold=0.0, include_par=False, female_threshold=0.2,
         Sex imputation statistics per sample.
-
     """
+    assert(aaf_threshold >= 0.0 and aaf_threshold <= 1.0)
```
Sorry, should have been clearer what I wanted. Something like this. Not sure what error to raise.
```
if not 0.0 <= aaf_threshold <= 1.0:
    raise FatalError("invalid argument for `aaf_threshold`. Must be in range [0, 1].")
```

tpoterba commented:
rebase conflict after merging splitmulti
